### PR TITLE
Firesim infrastructure updates, support for using most of the regular C++ STL in nanoPU, and nanoPU raft

### PR DESCRIPTION
**Pull request by amallery**
Fairly giant PR, sorry about that. Definitely continue resting instead of reading/merging it right away -- I just wanted to send it out now so that I'm not still sitting on all these changes. Theo, if you merge the latest changes from my branch (lnic-arg-passing) into your nuevomatch branch and make sure any of those conflicts are resolved, you should be able to merge your changes directly into the mainline lnic-dev branch after this PR merges in.

### CODE REVIEW

File: software/raft/raft_server_riscv/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+cmake_minimum_required(VERSION 2.8)
+
+project(raft_server_riscv)
+
+add_executable(raft_server_riscv ../../../tests/crt.o ../../../tests/syscalls.o main.cc mica/util/cityhash/city_mod.cc)
+target_link_libraries(raft_server_riscv raft -T /home/$ENV{USER}/chipyard/tests/link.ld -nostartfiles -lgcc -mcmodel=medany)
```
Comment (theojepsen) [inline software/raft/raft_server_riscv/CMakeLists.txt:6]: This assumes that the chipyard repo is always cloned into `$HOME`. Can we use a relative path instead like on the previous line? i.e. `../../../tests/link.ld`

File: software/raft/raft_server_riscv/firesim_riscv.cmake [hunk 2]
```diff
@@ -0,0 +1,13 @@
+set(CMAKE_SYSTEM_NAME Generic)
+set(CMAKE_SYSTEM_PROCESSOR riscv)
+set(tools /home/centos/chipyard/riscv-tools-install)
```
Comment (theojepsen) [inline software/raft/raft_server_riscv/firesim_riscv.cmake]: again, this assumes the repo is in `$HOME`

File: software/raft/raft_server_riscv/mica/util/cityhash/city.h [hunk 3]
```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2011 Google, Inc.
```
Comment (theojepsen) [inline software/raft/raft_server_riscv/mica/util/cityhash/city.h:1]: I think this file can be deleted. I originally imported these files from the eRPC repo. They made modifications to `city.{cc,h}` and renamed the files to `city_mod.{cc,h}`. Only one of these files is compiled and linked.

**Review by theojepsen — COMMENTED** [review]
I think you should
- remove `city.{h,cc}` because it's not compiled
- optionally, use relative paths instead of assuming chipyard is cloned into `$HOME`.

File: software/raft/raft_server_riscv/CMakeLists.txt (hunk 1 above)
Comment (sibanez12) [reply software/raft/raft_server_riscv/CMakeLists.txt:6]: Agreed. That would be good.

**Review by sibanez12 — COMMENTED** [review]
Super cool that you were able to figure out how to use STL with the nanoPU 😄 I'm sure this is going to be super helpful. Do you know if linking with STL will break or affect the performance of the ASPLOS tests (i.e. `lnic-evaluation.c`, `lnic-incast-single-core.c`, MICA, etc.)? Seeing as they don't use STL seems like they should be fine, but will the binary be much larger? Will we still be able to run Verilator tests?

File: software/raft/raft_server_riscv/mica/util/cityhash/city.h (hunk 3 above)
Comment (amallery) [reply software/raft/raft_server_riscv/mica/util/cityhash/city.h:1]: I was able to delete city.cc without any issues, but I think the existing code still depends on city.h via city_mod.h, so I'm keeping city.h for now.

File: software/raft/raft_server_riscv/firesim_riscv.cmake (hunk 2 above)
Comment (amallery) [reply software/raft/raft_server_riscv/firesim_riscv.cmake]: This entire file actually isn't needed any more, so I just took it out.

File: software/raft/raft_server_riscv/CMakeLists.txt (hunk 1 above)
Comment (amallery) [reply software/raft/raft_server_riscv/CMakeLists.txt:6]: Unfortunately, just changing this to a relative path causes the linker to fail to find the script. There might be a fancier cmake trick I can do to get the path though.

Comment (amallery) [reply software/raft/raft_server_riscv/CMakeLists.txt:6]: I fixed all of the other absolute paths I could find though.

**Comment by amallery** [comment]
Thanks! The test performance with STL included should be pretty much identical (although I haven't run it myself to verify), since the tests never call into the STL code at all. The binary might be a good bit bigger, however, which could cause some problems with running the tests in Verilator, since the serial program loader runs incredibly slowly in simulation and takes a very long time to load large binaries. It could be worth attempting to see how well it works -- if the linker is smart, it should notice that you're not actually using any STL functions and chop them out of the binary, which would be awesome.

**Review by sibanez12 — APPROVED** [review]
Thanks Alex. This looks good to me.

I haven't tested the ASPLOS tests with this STL linking either, but I'm ok to merge this in now and test it later when we go to make the ASPLOS results reproducible. We might need to update the tests/Makefile to avoid linking STL into the basic programs that don't need it, primarily to enable running fast Verilator sims. But let's cross that bridge when we get there.

**Review by theojepsen — APPROVED** [review]
:+1:  I think it's ready to be merged.

**Pull request merged**